### PR TITLE
Josec/codebook quantization

### PR DESCRIPTION
**Pull request by jecorona97**

### CODE REVIEW

File: tests/range_test.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+            [0.0, 0.0625, 0.125, 0.1875, 0.25, 0.3125, 0.375, 0.4375, 0.5, 0.5625,
+             0.625, 0.6875, 0.75, 0.8125, 0.875, 0.9375, 1.0, 1.0625, 1.125, 1.1875,
+             1.25, 1.3125, 1.375, 1.4375, 1.5, 1.5625, 1.625, 1.6875, 1.75, 1.8125,
+             1.875, 1.9375, -2.0, -1.9375, -1.875, -1.8125, -1.75, -1.6875, -1.625,
+             -1.5625, -1.5, -1.4375, -1.375, -1.3125, -1.25, -1.1875, -1.125, -1.0625,
+             -1.0, -0.9375, -0.875, -0.8125, -0.75, -0.6875, -0.625, -0.5625, -0.5,
+             -0.4375, -0.375, -0.3125, -0.25, -0.1875, -0.125, -0.0625]))
+    ])
+def test_quantized_bits_range(bits, integer, expected_values):
+  """Test quantized_relu range function."""
```
Comment (zhuangh) [inline tests/range_test.py:71]: """Test quantized_bits range function."""

File: qkeras/codebook.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+from __future__ import division
+from __future__ import print_function
+
+import warnings
+import numpy as np
+import tensorflow as tf
+import tensorflow.keras.backend as K
+from tensorflow.keras.layers import Layer, Embedding, Input
+from tensorflow.keras.models import Model
+from sklearn.cluster import KMeans
```
Comment (zhuangh) [inline qkeras/codebook.py:27]: please add sklearn>=$version in requirements.txt
also setup.py

File: qkeras/codebook.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    bits: Number of bits to compress weights to. This will
+      results in 2**bits codebook values
+    quantizer: quantizer function that will be applied to codebook values
+    rowwise: if true, rowwise clustering is applied. Otherwise Two-tier clustering
+
+  Returns:
+    km_models: list of fitted KMeans algorithms
+    quantized_embeddings: Numpy array MxN
+  """
+  from tqdm import tqdm
```
Comment (zhuangh) [inline qkeras/codebook.py:134]: add to requirements.txt and setup.py

File: qkeras/quantizers.py [hunk 4]
```diff
@@ -441,6 +440,18 @@ def min(self):
     else:
       return -1.0
 
+  def range(self):
+    """ Get the range of values that quantized_bits can represent """
```
Comment (zhuangh) [inline qkeras/quantizers.py:444]: the range of values? or Get the sorted values?

Comment (jecorona97) [reply qkeras/quantizers.py:444]: it currently outputs a list of all values it can output, unsorted. Should I sort them?

Comment (zhuangh) [reply qkeras/quantizers.py:444]: Yes. I think it is better to sort them for now. If we want to get the range, we could simply pick the first and last values.
Also change the docstring in this function, it is not "Get the range", it is "Gets the sorted values".

Comment (jecorona97) [reply qkeras/quantizers.py:444]: The thing is we the list of all possible values that can be outputted by the quantizer in order to create the in_table

File: qkeras/codebook.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+  Arguments:
+    weights: Numpy array
+    bits: Number of bits to compress weights to. This will
+      results in 2**bits codebook values
+    axis: axis to apply quantization by
+    quantizer: quantizer function that will be applied to codebook values
+
+  Returns:
+    km_models: list of fitted KMeans algorithms
+    quantized_embeddings: Numpy array MxN
```
Comment (zhuangh) [inline qkeras/codebook.py:133]: codebook_table:. ...

File: qkeras/codebook.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    if quantizer:
+      km.cluster_centers_ = quantizer(km.cluster_centers_).numpy()
+    km.cluster_centers_.sort(axis=0)
+
+    km_models[i] = km
+    codebook_table[i, :] = km.cluster_centers_.flatten()
+    preds = km.predict(w.reshape(-1, 1))
+    index_table.append(preds.reshape(original_shape))
+  
+  index_table = np.concatenate(index_table, axis)
```
Comment (zhuangh) [inline qkeras/codebook.py:155]: regarding index_table, is the shape of index_table the same as the shape of weight?

For example, a conv layer, we should have weights[kh, kw, k_in, k_out] approximates or equals codebook_table[axis][index_table[kh, kw, k_in, k_out]]

In your jupyter notebook, could you verify that new_conv_weights closes to/equals conv_weights around your code
```
new_conv_weights = np.zeros(conv_weights.shape)
for i in range(conv_weights.shape[axis]):
  new_conv_weights[:,:,:,i] = codebook_table[i][index_table[:,:,:,i]]
```

Comment (jecorona97) [reply qkeras/codebook.py:155]: 1. yes, index_table is the same shape of weights.
Here are the stats from the absolute difference from both weight matrices
"sp.stats.describe((abs(new_conv_weights - conv_weights)).ravel())"
DescribeResult(nobs=864, minmax=(1.3589859008789062e-05, 0.33330094814300537), mean=0.05314029095818653, variance=0.0026912063707205084, skewness=2.260220215254187, kurtosis=6.558495502282023)
2. After codebook quantization from 4 to 3 bits: val_acc: 0.6202 . Original quantized model val_acc: 0.6471.

Comment (zhuangh) [reply qkeras/codebook.py:155]: Hi Jose, if possible, could you write a codebook_test.py to assert those numbers? Thanks.

Comment (jecorona97) [reply qkeras/codebook.py:155]: Yeah I can add it tomorrow or Wednesday the latest.
